### PR TITLE
Remove UseJsonFallbackSerializer

### PR DESCRIPTION
**Pull request by ReubenBond**
Follow-on from #2193.

Json.NET is not a suitable fallback serializer. It does not handle the complex types which it needs to handle.

### CODE REVIEW

File: src/Orleans/Serialization/SerializationManager.cs [hunk 1]
```diff
@@ -143,36 +138,26 @@ public static object GetUninitializedObjectWithFormatterServices(Type type)
...
                 throw ex.Flatten();
             }
         }
 
-        internal static void Initialize(List<TypeInfo> serializationProviders, bool useJsonFallbackSerializer)
+        internal static void Initialize(List<TypeInfo> serializationProviders)
         {
             RegisterBuiltInSerializers();
-
-#if NETSTANDARD
```
Comment (sergeybykov) [inline src/Orleans/Serialization/SerializationManager.cs]: Don't we still need this #if and the warning on NETSTANDARD?

File: src/Orleans/Serialization/OrleansJsonSerializer.cs [hunk 2]
```diff
@@ -7,16 +7,20 @@
 
 namespace Orleans.Serialization
 {
-    internal class OrleansJsonSerializer : IExternalSerializer
+    using Orleans.Providers;
+
+    public class OrleansJsonSerializer : IExternalSerializer
```
Comment (sergeybykov) [inline src/Orleans/Serialization/OrleansJsonSerializer.cs]: Why do we need to make it public?

File: src/Orleans/Serialization/SerializationManager.cs (hunk 1 above)
Comment (ReubenBond) [inline src/Orleans/Serialization/SerializationManager.cs]: Eep, this entire block should be removed. Thanks for pointing this out!

File: src/Orleans/Serialization/SerializationManager.cs [hunk 3]
```diff
@@ -2210,40 +2186,5 @@ public SerializerMethods(DeepCopier deepCopy, Serializer serialize, Deserializer
             public Serializer Serialize { get; private set; }
             public Deserializer Deserialize { get; private set; }
         }
-
-        public static JsonSerializerSettings GetDefaultJsonSerializerSettings()
```
Comment (sergeybykov) [inline src/Orleans/Serialization/SerializationManager.cs]: Don't we still need this for other usages of `OrleansJsonSerializer`, e.g. by storage providers?

File: test/TesterInternal/Serialization/BuiltInSerializerTests.cs [hunk 4]
```diff
@@ -65,8 +59,7 @@ public BuiltInSerializerTests(ITestOutputHelper output)
         }
 
         [Theory, TestCategory("Functional"), TestCategory("Serialization")]
-        [InlineData(SerializerToUse.Default)]
```
Comment (sergeybykov) [inline test/TesterInternal/Serialization/BuiltInSerializerTests.cs]: What's this change for?

File: src/Orleans/Serialization/SerializationManager.cs (hunk 1 above)
Comment (ReubenBond) [inline src/Orleans/Serialization/SerializationManager.cs]: Oh, I misread the diff. I'm not sure that we need a warning which the user cannot do anything about. On the other hand, in `GetFallbackSerializer()`, I'm throwing unconditionally on `NETSTANDARD`, which is clearly untenable. What's the other option, though, other than to wait for #2162 for `NETSTANDARD` support.

We could return a fallback serializer which always throws.

File: src/Orleans/Serialization/OrleansJsonSerializer.cs [hunk 5]
```diff
@@ -7,16 +7,20 @@
...
         private static JsonSerializerSettings defaultSettings;
         private Logger logger;
-
+        
         /// <summary>
         /// Returns a configured <see cref="JsonSerializerSettings"/> 
         /// </summary>
         /// <returns></returns>
-        internal static JsonSerializerSettings GetDefaultSerializerSettings()
+        public static JsonSerializerSettings GetDefaultSerializerSettings()
```
Comment (sergeybykov) [inline src/Orleans/Serialization/OrleansJsonSerializer.cs]: Same question here.

Comment (ReubenBond) [inline src/Orleans/Serialization/OrleansJsonSerializer.cs]: This method was essentially moved from `SerializationManager`, but it's used by external providers in order to get serialization settings which are suitable for serializing types which we use (grain refs, etc).

File: test/TesterInternal/Serialization/BuiltInSerializerTests.cs (hunk 4 above)
Comment (ReubenBond) [inline test/TesterInternal/Serialization/BuiltInSerializerTests.cs]: It makes it easier to add & test other serializers (eg, `IlBasedSerializer`) without modifying every method.

In the (legitimate?) case where one serializer is intended to fail and another is intended to succeed, then we can use `InlineData` to specify that.

I'm happy to revert the change and simply remove the do-nothing `Fallback` option.

File: src/Orleans/Serialization/SerializationManager.cs (hunk 1 above)
Comment (sergeybykov) [inline src/Orleans/Serialization/SerializationManager.cs]: Maybe we should leave the warning here, and let `GetFallbackSerializer()` throw right after?

File: src/Orleans/Serialization/OrleansJsonSerializer.cs (hunk 2 above)
Comment (ReubenBond) [inline src/Orleans/Serialization/OrleansJsonSerializer.cs]: Users ought to be able to specify external serializers to use, even if only in certain cases.
This class could be used for storage providers, for example. Currently they use bits of this class (even if they were using it via `SerializationManager`).

File: src/Orleans/Serialization/SerializationManager.cs (hunk 3 above)
Comment (ReubenBond) [inline src/Orleans/Serialization/SerializationManager.cs]: See the other comment about moving the method. If you feel it's better to keep the methods on SerializationManager, then I can move them back. It seems messy, though.

A cleaner approach would be for those storage providers to consume an `IExternalSerializer` for all of their serialization, rather than each implementing support for JSON + Orleans serialization.

File: src/Orleans/Serialization/SerializationManager.cs (hunk 1 above)
Comment (ReubenBond) [inline src/Orleans/Serialization/SerializationManager.cs]: No problem, change incoming.

**Comment by ReubenBond** [comment]
> It appears as if there's more than just removing the option of using JSON as a fallback serializer here. But maybe I'm just confused.

You're right, I'll split the other changes into a separate PR. They were changes I made while I was cleaning up the JSON fallback serializer option, but they're not as targeted as the title suggests.

File: src/Orleans/Serialization/OrleansJsonSerializer.cs (hunk 5 above)
Comment (sergeybykov) [inline src/Orleans/Serialization/OrleansJsonSerializer.cs]: I don't disagree with your reasoning. But doesn't this mean that all storage providers that use `OrleansJsonSerializer` would have to be updated? If so, then I think it is better to make such a change explicit - as a separate PR, and probably targeting 1.4.0, since we are so close to releasing 1.3.0.

Comment (ReubenBond) [inline src/Orleans/Serialization/OrleansJsonSerializer.cs]: Agreed

**Comment by ReubenBond** [comment]
Updated - it's much slimmer now.
I retained the removal of the redundant `[InlineData(Fallback)]` attribute in `BuiltInSerializerTests`, since it previously had the same behavior as `JsonFallback`, which is now removed.
